### PR TITLE
fix(next-swc): Provide tokio context required for WASM plugins

### LINKED ISSUE #62081: Can't run application with any custom SWC plugin
### Link to the code that reproduces this issue

https://github.com/kevva/next-bugs/tree/swc-plugins

### To Reproduce

1. Run `npm run dev` (or `npm run build`) and you should see an error

### Current vs. Expected behavior

**Current behaviour**

When running the app using `next dev` or building using `next build` with a custom SWC plugin currently produces an error saying:

```
thread '<unnamed>' panicked at /Users/geist/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wasmer-wasix-0.18.0/src/runtime/task_manager/tokio.rs:105:19:
there is no reactor running, must be called from the context of a Tokio 1.x runtime
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Expected behaviour**

The application starts or builds successfully.

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.2.0: Wed Nov 15 21:53:18 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6000
Binaries:
  Node: 18.17.1
  npm: 10.2.4
  Yarn: 1.22.21
  pnpm: N/A
Relevant Packages:
  next: 14.1.1-canary.54 // Latest available version is detected (14.1.1-canary.54).
  eslint-config-next: 14.1.1-canary.54
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

SWC transpilation

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local)

### Additional context

This was introduced in https://github.com/vercel/next.js/releases/tag/v14.1.1-canary.52.

### PR DESCRIPTION
**Pull request by kdy1**
### What?

Run transforms with the tokio context so Wasmer works properly.

### Why?

To allow invoking Wasm plugins. `wasmer` had a breaking change.

### How?

Closes PACK-2587
Closes #62081

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/wasmer-tokio | Change |
| - | - | - | - |
| buildDuration | 13.9s | 14s | ⚠️ +150ms |
| buildDurationCached | 7.4s | 6.4s | N/A |
| nodeModulesSize | 197 MB | 197 MB | ✓ |
| nextStartRea..uration (ms) | 409ms | 409ms | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/wasmer-tokio | Change |
| - | - | - | - |
| 1068-HASH.js gzip | 30.3 kB | 30.3 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 4944-HASH.js gzip | 5.04 kB | 5.03 kB | N/A |
| 8423.HASH.js gzip | 181 B | 181 B | ✓ |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 241 B | 242 B | N/A |
| main-HASH.js gzip | 32.2 kB | 32.1 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 47.1 kB | 47.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/wasmer-tokio | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/wasmer-tokio | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 196 B | ✓ |
| _error-HASH.js gzip | 184 B | 183 B | N/A |
| amp-HASH.js gzip | 503 B | 504 B | N/A |
| css-HASH.js gzip | 323 B | 324 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.51 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 259 B | N/A |
| head-HASH.js gzip | 353 B | 351 B | N/A |
| hooks-HASH.js gzip | 370 B | 370 B | ✓ |
| image-HASH.js gzip | 4.21 kB | 4.2 kB | N/A |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.68 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 313 B | 314 B | N/A |
| script-HASH.js gzip | 386 B | 385 B | N/A |
| withRouter-HASH.js gzip | 309 B | 311 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 931 B | 931 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/wasmer-tokio | Change |
| - | - | - | - |
| _buildManifest.js gzip | 485 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/wasmer-tokio | Change |
| - | - | - | - |
| index.html gzip | 528 B | 527 B | N/A |
| link.html gzip | 540 B | 540 B | ✓ |
| withRouter.html gzip | 523 B | 522 B | N/A |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/wasmer-tokio | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94.9 kB | 94.9 kB | N/A |
| page.js gzip | 3.06 kB | 3.05 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/wasmer-tokio | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 625 B | 627 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.4 kB | 25.4 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/wasmer-tokio | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 170 kB | 170 kB | ✓ |
| app-page-exp..prod.js gzip | 96.5 kB | 96.5 kB | ✓ |
| app-page-tur..prod.js gzip | 98.3 kB | 98.3 kB | ✓ |
| app-page-tur..prod.js gzip | 92.8 kB | 92.8 kB | ✓ |
| app-page.run...dev.js gzip | 151 kB | 151 kB | ✓ |
| app-page.run..prod.js gzip | 91.3 kB | 91.3 kB | ✓ |
| app-route-ex...dev.js gzip | 22 kB | 22 kB | ✓ |
| app-route-ex..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 14.7 kB | 14.7 kB | ✓ |
| app-route.ru...dev.js gzip | 21.7 kB | 21.7 kB | ✓ |
| app-route.ru..prod.js gzip | 14.7 kB | 14.7 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.5 kB | 9.5 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.78 kB | 9.78 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.5 kB | 9.5 kB | ✓ |
| pages-turbo...prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| pages.runtim...dev.js gzip | 23 kB | 23 kB | ✓ |
| pages.runtim..prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| server.runti..prod.js gzip | 50.5 kB | 50.5 kB | ✓ |
| Overall change | 950 kB | 950 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js kdy1/wasmer-tokio | Change |
| - | - | - | - |
| 0.pack gzip | 1.55 MB | 1.55 MB | ⚠️ +143 B |
| index.pack gzip | 105 kB | 104 kB | N/A |
| Overall change | 1.55 MB | 1.55 MB | ⚠️ +143 B |

</details>

</details>
Commit: e32ab011271854ab413103072fe34d41c780d524

File: packages/next-swc/crates/napi/src/transform.rs [hunk 1]
```diff
@@ -71,90 +71,99 @@ impl Task for TransformTask {
...
+                                        Input::FromFilename => {
+                                            let filename = &options.swc.filename;
+                                            if filename.is_empty() {
+                                                bail!("no filename is provided via options");
+                                            }
+
+                                            self.c.cm.new_source_file(
+                                                FileName::Real(filename.into()),
+                                                read_to_string(filename).with_context(|| {
+                                                    format!(
```
Comment (ForsakenHarmony) [inline packages/next-swc/crates/napi/src/transform.rs:110]: this was already way too much indentation before and now it's even worse :/

can you extract parts of it into a different function or something like that?

File: packages/next-swc/crates/napi/src/transform.rs [hunk 2]
```diff
@@ -71,90 +71,99 @@ impl Task for TransformTask {
...
-                    |handler| {
-                        self.c.run(|| {
-                            let options: TransformOptions = serde_json::from_slice(&self.options)?;
-                            let fm = match &self.input {
-                                Input::Source { src } => {
-                                    let filename = if options.swc.filename.is_empty() {
-                                        FileName::Anon
-                                    } else {
-                                        FileName::Real(options.swc.filename.clone().into())
+        tokio::runtime::Runtime::new()
```
Comment (ForsakenHarmony) [inline packages/next-swc/crates/napi/src/transform.rs:74]: napi should already have a tokio runtime, can we use it here somehow?
https://napi.rs/docs/compat-mode/concepts/tokio

Comment (kdy1) [reply packages/next-swc/crates/napi/src/transform.rs]: We can't use it for `AsyncTask`

**Review by sokra — COMMENTED** [review]
napi provides a tokio runtime. Could we use that instead of creating a new one?

**Comment by ijjk** [comment]
## Failing test suites 
Commit: c7d85e865b0bd2cab92a3cf20b0aa4acdf3ffd00

`pnpm test-start test/e2e/app-dir/navigation/navigation.test.ts` (PPR)
- app dir - navigation > navigation between pages and app > should not continously initiate a mpa navigation to the same URL when router state changes

<details>
<summary>Expand output</summary>

  ● app dir - navigation › navigation between pages and app › should not continously initiate a mpa navigation to the same URL when router state changes

    page.waitForSelector: Timeout 60000ms exceeded.
    Call log:
      - waiting for locator('#link-to-slow-page')
      -   locator resolved to visible <a href="/slow-page" id="link-to-slow-page">To /slow-page</a>

      421 |     return this.chain(() => {
      422 |       return page
    > 423 |         .waitForSelector(selector, { timeout, state: 'attached' })
          |          ^
      424 |         .then(async (el) => {
      425 |           // it seems selenium waits longer and tests rely on this behavior
      426 |           // so we wait for the load event fire before returning

      at waitForSelector (lib/browsers/playwright.ts:423:10)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by sokra — APPROVED** [review]

File: packages/next-swc/crates/napi/src/transform.rs [hunk 3]
```diff
@@ -168,11 +175,11 @@ impl Task for TransformTask {
 
 #[napi]
 pub fn transform(
+    env: Env,
     src: Either3<String, Buffer, Undefined>,
     _is_module: bool,
     options: Buffer,
-    signal: Option<AbortSignal>,
-) -> napi::Result<AsyncTask<TransformTask>> {
+) -> napi::Result<Object> {
```
Comment (ForsakenHarmony) [inline packages/next-swc/crates/napi/src/transform.rs:182]: Will this act as a promise?

**Review by ForsakenHarmony — APPROVED** [review]

**Pull request merged**